cairo-libs/packages/protobuf/src/types/message.cairo

```diff
@@ -0,0 +1,169 @@
+                context2.buffer.len().encode_raw(ref self);
+            }
+            self.buffer.append(@context2.buffer);
+
+            i += 1;
+        }
+    }
+}
+
+#[derive(Drop)]
+pub struct DecodeContext {
+    pub buffer: @ByteArray,
+    pub index: usize,
+    pub limits: Array<usize>,
+}
+
+#[generate_trait]
+pub impl DecodeContextImpl of DecodeContextTrait {
+    fn new(buffer: @ByteArray) -> DecodeContext {
+        DecodeContext { buffer, index: 0, limits: array![] }
+    }
+
+    fn init_branch(ref self: DecodeContext, length: usize) {
+        let limit = self.index + length;
+        assert(limit <= self.buffer.len(), 'bigger limit for branch');
+        self.limits.append(self.index + length);
+    }
+
+    fn can_read_branch(ref self: DecodeContext) -> bool {
+        @self.index < self.limits[self.limits.len() - 1]
+    }
+
+    fn decode_field<T, +ProtoMessage<T>, +Drop<T>>(
+        ref self: DecodeContext, field_number: u8, ref value: T
+    ) {
+        if self.can_read_branch() {
+            let tag = ProtobufTagImpl::decode(self.buffer[self.index]);
+            if tag.field_number == field_number {
+                self.index += 1;
+
+                let wire_type = ProtoMessage::<T>::wire_type();
+
+                // println!(
+                //     "field_number: {}, actual_wire_type: {:?}, expected_wire_type: {:?}",
+                //     field_number,
+                //     tag.wire_type,
+                //     wire_type
+                // );
+
+                assert(wire_type == tag.wire_type, 'unexpected wire type');
+
+                if wire_type == WireType::LengthDelimited {
+                    let mut length = 0;
+                    length.decode_raw(ref self);
+                    self.init_branch(length);
+                    value.decode_raw(ref self);
+                    self.end_branch();
+                } else {
+                    value.decode_raw(ref self);
+                }
+            } else if tag.field_number < field_number {
+                panic!(
+                    "unexpected field number order: at expected field {} but got older field {}",
+                    field_number,
+                    tag.field_number
+                );
+            }
+        }
+    }
+
+    // for unpacked repeated fields (default for non-scalars)
+    fn decode_repeated_field<T, +ProtoMessage<T>, +Default<T>, +Drop<T>>(
+        ref self: DecodeContext, field_number: u8, ref value: Array<T>
+    ) {
+        while self.can_read_branch() {
+            let tag = ProtobufTagImpl::decode(self.buffer[self.index]);
+            if tag.field_number != field_number {
+                break;
+            }
+            let mut item = Default::<T>::default();
+
+            self.decode_field(field_number, ref item);
+            value.append(item);
+        };
+    }
+
+    fn end_branch(ref self: DecodeContext) {
+        // TODO(rano): pop_back is not impl for Array<T>, this is inefficient
```

Comment:
Do we have a stack implementation in Cairo ? It seems, the `Array` is actually a queue implementation.